File: python/25_mps2statevector.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Generate statevector from MPS
"""

import numpy as np


def main():
    # Bell state
    print("Bell state:")
    tl = np.array([[1, 0], [0, 1]]) / np.pow(2, 0.25)
    tr = np.array([[1, 0], [0, 1]]) / np.pow(2, 0.25)
    print(f"left tensor:\n{tl}")
    print(f"right tensor:\n{tr}\n")

    bell = np.einsum("ij,jk->ik", tl, tr)
    bell = bell.reshape(-1)
    print(f"statevector:\n{bell}\n")

    # GHZ state
    n = 6
    print(f"n={n} GHZ state:")
    w = np.pow(2, 1 / (2 * n))
    tl = np.array([[1, 0], [0, 1]]) / w
    tr = np.array([[1, 0], [0, 1]]) / w
    t = np.zeros((2, 2, 2))
    t[0, 0, 0] = 1 / w
    t[1, 1, 1] = 1 / w
    print(f"left tensor:\n{tl}")
    print(f"right tensor:\n{tr}")
    print(f"middle tensors:\n{t}\n")

    ghz = tl
    for _ in range(1, n - 1):
        ghz = np.einsum("ij,jkl->ikl", ghz, t)
        ghz = ghz.reshape(ghz.shape[0] * ghz.shape[1], ghz.shape[2])
    ghz = np.einsum("ij,jk->ik", ghz, tr)
    ghz = ghz.reshape(-1)
    print(f"statevector:\n{ghz}")


if __name__ == "__main__":
    main()

File: julia/25_mps2statevector.jl
#!/usr/bin/env julia
"""
Generate statevector from MPS
"""

using TensorOperations

function main()
    # Bell state
    println("Bell state:")
    tl = [1.0 0.0; 0.0 1.0] / 2^(0.25)
    tr = [1.0 0.0; 0.0 1.0] / 2^(0.25)
    println("left tensor: ", tl)
    println("right tensor: ", tr, "\n")

    @tensor bell[i, k] := tl[i, j] * tr[j, k]
    bell = vec(permutedims(bell))
    println("statevector: ", bell, "\n")

    # GHZ state
    n = 6
    println("n=$(n) GHZ state:")
    w = 2^(1 / (2n))
    tl = [1.0 0.0; 0.0 1.0] / w
    tr = [1.0 0.0; 0.0 1.0] / w
    t = zeros(Float64, 2, 2, 2)
    t[1, 1, 1] = 1 / w
    t[2, 2, 2] = 1 / w
    println("left tensor: ", tl)
    println("right tensor: ", tr)
    println("middle tensors: ", t, "\n")

    ghz = tl
    for _ = 2:(n-1)
        @tensor tmp[i, k, l] := ghz[i, j] * t[j, k, l]
        ghz = reshape(tmp, size(tmp, 1) * size(tmp, 2), size(tmp, 3)) # (i*k, l)
    end
    @tensor ghz[i, k] := ghz[i, j] * tr[j, k]
    ghz = vec(permutedims(ghz))
    println("statevector: ", ghz)
end

main()
